### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by dhrdh**

### CODE REVIEW

File: usecase/demo.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
+package usecase
+
+import (
+	"Go-Dispatch-Bootcamp/types"
```
Comment (MigGrover) [inline usecase/demo.go:4]: It is a good practice to have the packages in the following order
``` go
import (
               "core packages"

               "your packages"

              "third party packages"
)
```

File: usecase/demo.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	GetFeedUsers(string) ([][]string, error)
+	FetchCsvFromRemote(string) ([][]string, error)
+	UpdateUsers(*[]types.User, string) (bool, error)
+}
+
+type demoUsecase struct {
+	service demoService
+}
+
+func New(s demoService) *demoUsecase {
```
Comment (MigGrover) [inline usecase/demo.go:26]: It is a good practice to comment on functions, interfaces and structures that are being exported.

File: usecase/demo.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	return &result, nil
+}
+
+func (tu *demoUsecase) Feed() ([][]string, error) {
+	log.Println("In usecase | Feed")
+
+	return tu.service.GetFeedUsers(feedFileName)
+}
+
+func (tu *demoUsecase) UpdateUsersFromFeed() (bool, error) {
```
Comment (MigGrover) [inline usecase/demo.go:70]: Set comment for exported function

File: controller/demo.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
+package controller
+
+import (
```
Comment (MigGrover) [inline controller/demo.go:3]: Remember it is a good practice to have packages in order for visibility

``` go
import (
               "core packages"

               "your packages"

               "third party packages"
)

File: controller/demo.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	FetchById(int) (*types.User, error)
+	Feed() ([][]string, error)
+	UpdateUsersFromFeed() (bool, error)
+}
+
+type demoController struct {
+	usecase usecase
+}
+
+func (tc *demoController) Fetch(w http.ResponseWriter, r *http.Request) {
```
Comment (MigGrover) [inline controller/demo.go:27]: It is a good practice to set comments

File: controller/demo.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	w.WriteHeader(http.StatusOK)
+	w.Header().Set("Content-Type", "application/json")
+
+	data, _ := json.Marshal(users)
+	w.Write(data)
+}
+
+func (tc *demoController) FetchById(w http.ResponseWriter, r *http.Request) {
```
Comment (MigGrover) [inline controller/demo.go:45]: Set comments for exported objects

File: controller/demo.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	w.WriteHeader(http.StatusOK)
+	w.Header().Set("Content-Type", "application/json")
+
+	data, _ := json.Marshal(user)
+	w.Write(data)
+}
+
+func (tc *demoController) Feed(w http.ResponseWriter, r *http.Request) {
```
Comment (MigGrover) [inline controller/demo.go:79]: Set comments for exported function

File: controller/demo.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	w.WriteHeader(http.StatusOK)
+	b := new(bytes.Buffer)
+	csvWriter := csv.NewWriter(b)
+	csvWriter.WriteAll(users)
+	w.Write(b.Bytes())
+}
+
+func (tc *demoController) UpdateUsersFromFeed(w http.ResponseWriter, r *http.Request) {
```
Comment (MigGrover) [inline controller/demo.go:97]: Set comments for exported objects

File: main.go [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+	"Go-Dispatch-Bootcamp/usecase"
+)
+
+func main() {
+	demoService := service.New()
+	demoUsecase := usecase.New(demoService)
+	demoController := controller.New(demoUsecase)
+	httpRouter := router.Setup(demoController)
+
+	err := http.ListenAndServe("localhost:8080", httpRouter)
```
Comment (MigGrover) [inline main.go:19]: It would be great if you could obtain this from a configure file or environment configs as two parts the port and the host.

Comment (MigGrover) [reply main.go:19]: or even separate it to a constant would give a better visibility :)

File: router/demo.go [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+)
+
+type demoController interface {
+	Fetch(w http.ResponseWriter, r *http.Request)
+	FetchById(w http.ResponseWriter, r *http.Request)
+	Feed(w http.ResponseWriter, r *http.Request)
+	UpdateUsersFromFeed(w http.ResponseWriter, r *http.Request)
+}
+
+func Setup(c demoController) *mux.Router {
```
Comment (MigGrover) [inline router/demo.go:17]: Set some comments, about exported functions, this is a good practice and is recommended so whenever other people would try to add new code or modify the current one, could get a context without checking all the project.

// Setup do..., expect... and will return... would be good enough but there are a lot of styles for this comments

File: service/demo.go [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+	}
+	reader := csv.NewReader(file)
+	records, err := reader.ReadAll()
+	if err != nil {
+		return nil, errors.New("read file error")
+	}
+	return records, nil
+}
+
+func (ts *demoService) FetchCsvFromRemote(feedUrl string) ([][]string, error) {
```
Comment (MigGrover) [inline service/demo.go:38]: remember to set comments for exported functions

File: service/demo.go [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+	reader.Comma = ','
+	data, err := reader.ReadAll()
+	if err != nil {
+		return nil, errors.New("can not read csv from remote")
+	}
+
+	return data, nil
+}
+
+func (ts *demoService) UpdateUsers(users *[]types.User, dataFileName string) (bool, error) {
```
Comment (MigGrover) [inline service/demo.go:59]: remember to set comments for exported functions

File: service/demo.go [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+	if err != nil {
+		return false, err
+	}
+
+	log.Println(fmt.Sprintf("%v was updated", dataFileName))
+
+	return true, nil
+}
+
+func (ts *demoService) GetUsers(dataFileName string) (*[]types.User, error) {
```
Comment (MigGrover) [inline service/demo.go:94]: remember to set comments for exported functions

File: service/demo.go [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+			Identifier: line[2],
+			FirstName:  line[3],
+			LastName:   line[4],
+		})
+	}
+
+	return &users, nil
+}
+
+func (ts *demoService) GetUsersMap(dataFileName string) (map[int]types.User, error) {
```
Comment (MigGrover) [inline service/demo.go:123]: remember to set comments for exported functions

File: service/demo.go [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+			Identifier: line[2],
+			FirstName:  line[3],
+			LastName:   line[4],
+		}
+	}
+
+	return users, nil
+}
+
+func (ts *demoService) GetFeedUsers(feedFileName string) ([][]string, error) {
```
Comment (MigGrover) [inline service/demo.go:152]: remember to set comments for exported functions

File: service/demo_test.go [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+			result, err := tt.ts.GetFeedUsers(feedFileName)
+
+			assert.Equal(t, tt.want, result)
+			assert.NoError(t, err)
+		})
+	}
+}
+
+// To make this test pass you need to run the server
+func Test_demoService_FetchCsvFromRemote(t *testing.T) {
```
Comment (MigGrover) [inline service/demo_test.go:191]: This should not happen, you could simulate a server if you are expecting a response, there are ways to create a mock server using Go. I would suggest httptest for this.

**Review by MigGrover — COMMENTED** [review]
Here are three points:

1. Not totally sure about log.Println("In...") be careful as this would generate a lot of data for your log each time a request is being done.
2. Be careful and remember to ask questions whenever there are doubts, You are consuming your own CSV, and I'm not sure if this is because you did not find an external API that adjusted to your needs or because there was a misunderstanding about how does the project work.
3. The name of your objects should be coherent with the entity you are passing, having all except by the entity being called "demo" and the entity User, was kind of confusing.

File: usecase/demo.go (hunk 1 above)
Comment (dhrdh) [reply usecase/demo.go:4]: Trying to do like this. thx

File: usecase/demo.go (hunk 2 above)
Comment (dhrdh) [reply usecase/demo.go:26]: ok

File: usecase/demo.go (hunk 3 above)
Comment (dhrdh) [reply usecase/demo.go:70]: ok

File: controller/demo.go (hunk 4 above)
Comment (dhrdh) [reply controller/demo.go:3]: Trying to do like this. thx

File: controller/demo.go (hunk 5 above)
Comment (dhrdh) [reply controller/demo.go:27]: Trying to do like this. thx

File: service/demo_test.go (hunk 16 above)
Comment (dhrdh) [reply service/demo_test.go:191]: ok will try later on

**Pull request merged**